strum_macros/src/macros/enum_table.rs

```diff
@@ -0,0 +1,193 @@
+        transform_fields.push(quote! {#snake_case: func(#name::#pascal_case, &self.#snake_case),});
+        pascal_idents.push(pascal_case);
+        snake_idents.push(snake_case);
+    }
+
+    // Error on empty enums
+    if pascal_idents.is_empty() {
+        return Err(syn::Error::new(
+            variants.span(),
+            "`EnumTable` requires at least one non-disabled variant",
+        ));
+    }
+
+    // if the index operation can panic, add that to the documentation
+    if !disabled_variants.is_empty() {
+        doc_comment.push_str(&format!(
+            "\n# Panics\nIndexing `{table_name}` with any of the following variants will cause a panic:"
+        ));
+        for variant in disabled_variants {
+            doc_comment.push_str(&format!("\n\n- `{name}::{variant}`"));
+        }
+    }
+
+    let doc_new = format!("Create a new {table_name} with a value for each variant of {name}");
+    let doc_closure =
+        format!("Create a new {table_name} by running a function on each variant of `{name}`");
+    let doc_transform = format!("Create a new `{table_name}` by running a function on each variant of `{name}` and the corresponding value in the current `{table_name}`");
+    let doc_filled = format!("Create a new `{table_name}` with the same value in each field.");
+    let doc_option_all = format!("Converts `{table_name}<Option<T>>` into `Option<{table_name}<T>>`. Returns `Some` if all fields are `Some`, otherwise returns `None`.");
+    let doc_result_all_ok = format!("Converts `{table_name}<Result<T, E>>` into `Result<{table_name}, E>`. Returns `Ok` if all fields are `Ok`, otherwise returns `Err`.");
+
+    Ok(quote! {
+        #[doc = #doc_comment]
+        #[allow(
+            missing_copy_implementations,
+        )]
+        #[derive(Debug, Clone, Default, PartialEq, Eq, Hash)]
+        #vis struct #table_name<T> {
+            #(#snake_idents: T,)*
+        }
+
+        impl<T: Clone> #table_name<T> {
+            #[doc = #doc_filled]
+            #vis fn filled(value: T) -> #table_name<T> {
+                #table_name {
+                    #(#snake_idents: value.clone(),)*
+                }
+            }
+        }
+
+        impl<T> #table_name<T> {
+            #[doc = #doc_new]
+            #vis fn new(
+                #(#snake_idents: T,)*
+            ) -> #table_name<T> {
+                #table_name {
+                    #(#snake_idents,)*
+                }
+            }
+
+            #[doc = #doc_closure]
+            #vis fn from_closure<F: Fn(#name)->T>(func: F) -> #table_name<T> {
+              #table_name {
+                #(#closure_fields)*
+              }
+            }
+
+            #[doc = #doc_transform]
+            #vis fn transform<U, F: Fn(#name, &T)->U>(&self, func: F) -> #table_name<U> {
+              #table_name {
+                #(#transform_fields)*
+              }
+            }
+
+        }
+
+        impl<T> core::ops::Index<#name> for #table_name<T> {
```

Comment:
I suggest doing `::core::ops::Index` to ensure we always pull this method from `core` in case someone shadowed the module.